### PR TITLE
feat: Implement 'and' intrinsic function

### LINKED ISSUE #4380: enh: Support logical kind args in `and`, `or` and `xor`
```fortran
program main
    print *, and(.true., .false.)
end program main
```
```console
 lfortran a.f90
semantic error: Unexpected args, Iand expects (int, int) as arguments
 --> a.f90:2:14
  |
2 |     print *, and(.true., .false.)
  |              ^^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
$ gfortran a.f90 && ./a.out
 F
```

We also need to support `boz-literal-constant` args too.

### PR DESCRIPTION
**Pull request by jinangshah21**
Fixes #4380. 
Still not Completed(Draft).

### CODE REVIEW

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
```diff
diff --git a/src/libasr/intrinsic_func_registry_util_gen.py b/src/libasr/intrinsic_func_registry_util_gen.py
index 6a4cd2de7..9718cf537 100644
--- a/src/libasr/intrinsic_func_registry_util_gen.py
+++ b/src/libasr/intrinsic_func_registry_util_gen.py
@@ -577,7 +577,7 @@ intrinsic_funcs_args = {
     ],
     "Iand": [
         {
-            "args": [("int", "int")],
+            "args": [("int", "int"), ("bool", "bool")],
             "ret_type_arg_idx": 0,
             "same_kind_arg": 2
         },
```

As suggested by @HarshitaKalani, `Iand` only supports integer argument. Let's create a new intrinsic `And` which allows `integer` and `boolean`. Make frontend recognise it.

**Comment by jinangshah21** [comment]
Once this get approved we can do same for 'or' and 'xor'.

File: src/libasr/intrinsic_func_registry_util_gen.py [hunk 1]
```diff
@@ -582,6 +582,13 @@
             "same_kind_arg": 2
         },
     ],
+    "And": [
+        {
+            "args": [("int", "int"),("bool","bool")],
+            "ret_type_arg_idx": 0,
+            "same_kind_arg": 2
```
[the remark this answers is not in the document]
Comment (jinangshah21) [reply src/libasr/intrinsic_func_registry_util_gen.py:589]: Yes they can be of different kind, I will remove that line.

**Review by HarshitaKalani — APPROVED** [review]
This looks good to me. I think we can merge this.
And then think about the idea mentioned above.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -921,7 +922,6 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         {"ddim", {"dim", {"real8", "real8"}}},
         {"amod", {"mod", {"real4", "real4"}}},
         {"dmod", {"mod", {"real8", "real8"}}},
-        {"and", {"iand", {"any", "any"}}},
```
[the remark this answers is not in the document]
Comment (jinangshah21) [reply src/lfortran/semantics/ast_common_visitor.h:924]: > I can think of a better approach here. We can keep the `and` implementation and remove the `iand` implementation and then map it this way:
> 
> ```
> {"iand", {"and", {"int", "int"}}},
> ```
> 
> How does that sound to you?

There is an issue doing this : `and` accepts only scalar arguments but `iand` can have an array as input as per GFortran.

[the remark this answers is not in the document]
Comment (HarshitaKalani) [reply src/lfortran/semantics/ast_common_visitor.h:924]: I see. Then we'll need to see the correct way of implementing `and` as currently it is implemented as an intrinsic elemental function with this PR.

[the remark this answers is not in the document]
Comment (jinangshah21) [reply src/lfortran/semantics/ast_common_visitor.h:924]: Can we keep a flag "scalar" in `intrinsic_function_registry_gen_util.py` that can check if input is scalar Or not?

**Pull request merged**